File: BSharp/Controllers/DTO/MeasurementUnit.cs
using BSharp.Controllers.Misc;
using Microsoft.Extensions.Localization;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace BSharp.Controllers.DTO
{
    /// <summary>
    /// All savable DTOs must inherit from <see cref="DtoForSaveKeyBase{TKey}"/>
    /// </summary>
    [StrongEntity]
    public class MeasurementUnitForSave : DtoForSaveKeyBase<int?>
    {
        [BasicField]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [MultilingualDisplay(Name = "Name", Language = Language.Primary)]
        public string Name { get; set; }

        [BasicField]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [MultilingualDisplay(Name = "Name", Language =  Language.Secondary)]
        public string Name2 { get; set; }

        [BasicField]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [Display(Name = "Code")]
        public string Code { get; set; }

        [ChoiceList(new object[] { "Pure", "Time", "Distance", "Count", "Mass", "Volume", "Money" },
            new string[] { "MU_Pure", "MU_Time", "MU_Distance", "MU_Count", "MU_Mass", "MU_Volume", "MU_Money" })]
        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [StringLength(255, ErrorMessage = nameof(StringLengthAttribute))]
        [Display(Name = "MU_UnitType")]
        public string UnitType { get; set; }

        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [Display(Name = "MU_UnitAmount")]
        public double? UnitAmount { get; set; }

        [Required(ErrorMessage = nameof(RequiredAttribute))]
        [Display(Name = "MU_BaseAmount")]
        public double? BaseAmount { get; set; }
    }

    /// <summary>
    /// The read-DTO, which always inherits from the update-DTO
    /// </summary>
    public class MeasurementUnit : MeasurementUnitForSave
    {
        [BasicField]
        [Display(Name = "IsActive")]
        public bool? IsActive { get; set; }

        [Display(Name = "CreatedAt")]
        public DateTimeOffset? CreatedAt { get; set; }

        [ForeignKey]
        [Display(Name = "CreatedBy")]
        public int? CreatedById { get; set; }

        [Display(Name = "ModifiedAt")]
        public DateTimeOffset? ModifiedAt { get; set; }

        [ForeignKey]
        [Display(Name = "ModifiedBy")]
        public int? ModifiedById { get; set; }

        // For Query

        [NavigationProperty(ForeignKey = nameof(CreatedById))]
        public LocalUser CreatedBy { get; set; }

        [NavigationProperty(ForeignKey = nameof(ModifiedById))]
        public LocalUser ModifiedBy { get; set; }
    }
}

File: BSharp/ClientApp/src/app/data/dto/measurement-unit.ts
import { DtoForSaveKeyBase } from './dto-for-save-key-base';

export class MeasurementUnitForSave extends DtoForSaveKeyBase {
  Name: string;
  Name2: string;
  Code: string;
  UnitType: 'Pure' | 'Time' | 'Distance' | 'Count' | 'Mass' | 'Volume' | 'Money';
  UnitAmount: number;
  BaseAmount: number;
}

export class MeasurementUnit extends MeasurementUnitForSave {
  IsActive: boolean;
  CreatedAt: string;
  CreatedBy: string;
  ModifiedAt: string;
  ModifiedBy: string;
}

// Choice list (Also repeated in measurement units master template)
export const MeasurementUnit_UnitType = {
  'Pure': 'MU_Pure',
  'Time': 'MU_Time',
  'Distance': 'MU_Distance',
  'Count': 'MU_Count',
  'Mass': 'MU_Mass',
  'Volume': 'MU_Volume',
  'Money': 'MU_Money'
};
